### PR TITLE
feat: add support for discussion grading

### PR DESCRIPTION
**Pull request by BryanttV**
### Description

This PR adds support for discussion grading. Instructors can set up a **Discussions Grading** component and learners can receive a grade depending on their participation in the discussion forum. The following fields were included in the component configuration:

* **Grading Method**: Allows to configure the grading method.
* **Maximum Attempts**: Allows to set the maximum number of attempts the learner can calculate their grade.
* **Number of Participations**: Allows to configure the number of participations the learner should make in the discussion forum.
* **Problem Weight**: Allows to define the number of points the problem is worth.
* **Instructions Text**: Allows to configure the text to be displayed to the learner with instructions for using the component.
* **Button Text**: Allows to configure the button's text for calculating the grade.

### How to Test
1. Install this branch.
2. Add `discussion_grading` to **Studio** > **Course** > **Settings** > **Advanced Settings**.
3. Add a new component of **Discussion Grading** in a unit and configure it.
4. Enables discussions using [tutor-forum](https://github.com/overhangio/tutor-forum/?tab=readme-ov-file#installation) plugin.
5. From the LMS in the discussions tab, as a learner create threads, posts and replies.
6. Go to the unit with the **Discussion Grading** component, and press the button. You can see your score depending of your discussion forum participation.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -72,7 +72,7 @@ insert_final_newline = true
 charset = utf-8
 indent_style = space
 indent_size = 4
-max_line_length = 120
+max_line_length = 100
 trim_trailing_whitespace = true
 
 [{Makefile, *.mk}]
```
Comment (mariajgrimaldi) [inline .editorconfig]: Is this file needed? Also, we usually use 120 in the other repos.

File: Makefile [hunk 2]
```diff
@@ -15,7 +15,7 @@ EXTRACTED_TEXT := $(EXTRACT_DIR)/text.po
 JS_TARGET := $(PACKAGE_NAME)/public/js/translations
 TRANSLATIONS_DIR := $(PACKAGE_NAME)/translations
 SOURCES=./setup.py ./$(PACKAGE_NAME)
-BLACK_OPTS = --exclude ${SOURCES}
+BLACK_OPTS = --exclude ${SOURCES} --line-length 100
```
Comment (mariajgrimaldi) [inline Makefile]: Same comment here, we usually use 120. Docs: https://docs.openedx.org/en/latest/developers/references/developer_guide/style_guides/python-guidelines.html#syntax-and-organization

File: discussion_grading/discussion_grading.py [hunk 3]
```diff
@@ -2,33 +2,142 @@
...
+
+    grading_method = String(
+        display_name=_("Grading Method"),
+        help=_(
+            "Defines the grading method to be used. When set to 'Minimum "
+            "Participations', the learner will obtain the maximum score if the "
+            "number of participations is greater than or equal to the number of "
+            "participations required to pass. When set to 'Average Participations', "
+            "the learner will obtain a score equal to the number of participations "
+            "divided by the number of participations required to pass."
```
Comment (mariajgrimaldi) [inline discussion_grading/discussion_grading.py]:
```suggestion
        help=_(
            "Defines the grading method to be used. When set to 'Minimum "
            "Participations', the learner will obtain the maximum score if the "
            "number of participations is greater than or equal to the number of "
            "participations required to pass, 0 otherwise. When set to 'Average Participations', "
            "the learner will obtain a score equal to the number of participations "
            "divided by the number of participations required to pass."
```

File: discussion_grading/discussion_grading.py [hunk 4]
```diff
@@ -44,35 +153,223 @@ def resource_string(self, path: str) -> str:
...
+        Args:
+            user_stats (dict): The number of participations for the current user.
+
+        Returns:
+            int: The grade for the current user.
+        """
+        number_of_participations = sum(user_stats.values())
+
+        if number_of_participations >= self.number_of_participations:
+            return 1
```
Comment (mariajgrimaldi) [inline discussion_grading/discussion_grading.py]: Could this be a constant instead of a hardcoded number?

File: discussion_grading/discussion_grading.py [hunk 5]
```diff
@@ -44,35 +153,223 @@ def resource_string(self, path: str) -> str:
...
+
+        if self.grading_method == DiscussionGradingMethod.MINIMUM_PARTICIPATIONS.name:
+            return int(number_of_participations >= self.number_of_participations)
+        elif self.grading_method == DiscussionGradingMethod.AVERAGE_PARTICIPATIONS.name:
+            return number_of_participations / self.number_of_participations
```
Comment (mariajgrimaldi) [inline discussion_grading/discussion_grading.py:293]: I thought the submissions API didn't accept non-integer values. I don't quite remember, though.

File: discussion_grading/discussion_grading.py [hunk 6]
```diff
@@ -44,35 +153,223 @@ def resource_string(self, path: str) -> str:
...
+        Returns:
+            dict: The user stats for the current user or a message.
+        """
+        try:
+            user_stats = get_course_user_stats(self.block_course_id).get("user_stats")
+        except Exception:  # pylint: disable=broad-except
+            return {
+                "sucess": False,
+                "message": _("Discussion forum is not enabled."),
+            }
```
Comment (mariajgrimaldi) [inline discussion_grading/discussion_grading.py]: If we know that this exception means that the discussion forum is not enabled, why don't we catch that exception instead of Exception?

File: discussion_grading/discussion_grading.py [hunk 7]
```diff
@@ -44,35 +153,223 @@ def resource_string(self, path: str) -> str:
...
+        for user_stat in user_stats:
+            if user_stat.get("username") == get_username(self.current_user):
+                return {
+                    "sucess": True,
+                    "user_stats": {
+                        "threads": user_stat.get("threads"),
+                        "responses": user_stat.get("responses"),
+                        "replies": user_stat.get("replies"),
+                    },
+                }
```
Comment (mariajgrimaldi) [inline discussion_grading/discussion_grading.py]:
```suggestion
        username = get_username(self.current_user):
        for user_stat in user_stats:
            if user_stat.get("username") == username:
                return {
                    "sucess": True,
                    "user_stats": {
                        "threads": user_stat.get("threads"),
                        "responses": user_stat.get("responses"),
                        "replies": user_stat.get("replies"),
                    },
                }
```

File: discussion_grading/discussion_grading.py [hunk 8]
```diff
@@ -44,35 +153,223 @@ def resource_string(self, path: str) -> str:
...
+        if self.max_attempts and self.attempts >= self.max_attempts:
+            return {
+                "success": False,
+                "message": _("You have reached the maximum number of attempts."),
+            }
+
+        user_stats = self.get_user_stats()
+
+        if not user_stats["sucess"]:
+            return user_stats
```
Comment (mariajgrimaldi) [inline discussion_grading/discussion_grading.py]: What do you think about returning the error here instead of get_user_stats(), since this is the handler itself.

File: .editorconfig (hunk 1 above)
Comment (BryanttV) [reply .editorconfig]: Yes, the file is required.

File: Makefile (hunk 2 above)
Comment (BryanttV) [reply Makefile]: Yes, for consistency I will leave 120 as a limit.

Comment (BryanttV) [reply Makefile]: Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/b7fbbc27a38d4555abbdab7817a87c7081ce496c

File: discussion_grading/discussion_grading.py (hunk 3 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py]: Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/28dafb2282a1f60e486affd6ccc8aa62d2246d0c

File: discussion_grading/discussion_grading.py (hunk 4 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py]: Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/efbe6d40b7d8e569dd544b1a2370e194fa5c74df

File: discussion_grading/discussion_grading.py (hunk 5 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py:293]: Yes, that's why we use the `round` function when calling `set_score`

File: discussion_grading/discussion_grading.py (hunk 7 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py]: Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/74d1e9a800f34605ca95bb00a927fccf01b89580

File: discussion_grading/discussion_grading.py (hunk 6 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py]: The exception that returns is specific to the `requests` library. For that we should install that library, or I'm not sure how else to do it.

File: discussion_grading/discussion_grading.py (hunk 8 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py]: Yes, I agree
Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/c8b8783012e9449883e420262590e37d5868e478

File: discussion_grading/discussion_grading.py (hunk 6 above)
Comment (mariajgrimaldi) [reply discussion_grading/discussion_grading.py]: What if the call raises other exceptions, and we're catching it here with `Discussion forum is not enabled.` but that's not the error? In the long term, catching a specific exception type is better, or we could also use a exception-specific message like `exc.message` instead.

File: discussion_grading/discussion_grading.py (hunk 5 above)
Comment (mariajgrimaldi) [reply discussion_grading/discussion_grading.py:293]: Should the responsibility of returning the score already rounded lay on this method instead?

File: discussion_grading/constants.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+"""This module contains the constants used in the Discussion Grading XBlock."""
```
Comment (mariajgrimaldi) [inline discussion_grading/constants.py]: Why do we need a new module for constants if they're used in just one module? Could we consider adding them in the same module? Let me know.

File: discussion_grading/discussion_grading.py (hunk 6 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py]: Then, do you see fit to install the library?

File: discussion_grading/constants.py (hunk 9 above)
Comment (BryanttV) [reply discussion_grading/constants.py]: Yes, you are right
Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/79799038af7d261558e80dafebb79139f114f433

File: discussion_grading/discussion_grading.py (hunk 5 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py:293]: This must be done in the `set_score`, since it is multiplied by the weight.

File: discussion_grading/discussion_grading.py [hunk 10]
```diff
@@ -2,33 +2,147 @@
...
+loader = ResourceLoader(__name__)
 
+ITEM_TYPE = "discussion_grading"
+MAX_SCORE = 1
+MIN_SCORE = 0
 
-class XBlockDiscussionGrading(StudioEditableXBlockMixin, XBlock):
+
+@XBlock.needs("i18n")
+@XBlock.needs("user")
```
Comment (mariajgrimaldi) [inline discussion_grading/discussion_grading.py]: I just found out that you can do:
```suggestion
@XBlock.needs("i18n", "user")
```

File: discussion_grading/discussion_grading.py [hunk 11]
```diff
@@ -2,33 +2,147 @@
...
+    grading_method = String(
+        display_name=_("Grading Method"),
+        help=_(
+            "Defines the grading method to be used. When set to 'Minimum "
+            "Participations', the learner will obtain the maximum score if the "
+            "number of participations is greater than or equal to the number of "
+            "participations required to pass, 0 otherwise. When set to 'Average "
+            "Participations', the learner will obtain a score equal to the number "
+            "of participations divided by the number of participations required "
+            "to pass."
```
Comment (mariajgrimaldi) [inline discussion_grading/discussion_grading.py:70]: Let's put a pin here: we should add how it rounds (to the closest integer, ...) depending on what the product owner says

**Review by mariajgrimaldi — APPROVED** [review]
thank you!

File: discussion_grading/enums.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+class DiscussionGradingMethod(Enum):
+    """
+    Enum for discussion grading method.
+
+    - MINIMUM_PARTICIPATIONS: Learners are graded based on the minimum number of participations
+    - AVERAGE_PARTICIPATIONS: Learners are graded based on the average number of participations
+    """
+
+    MINIMUM_PARTICIPATIONS = _("Minimum Participations")
+    AVERAGE_PARTICIPATIONS = _("Average Participations")
```
Comment (mariajgrimaldi) [inline discussion_grading/enums.py]: Same suggestion here about having a separate file for enums. Should we include them in the utils.py file or in the discussion_grading.py file?

File: discussion_grading/discussion_grading.py (hunk 10 above)
Comment (BryanttV) [reply discussion_grading/discussion_grading.py]: Wow, I didn't know you could do that 
Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/06352b709b38413ad582d918283eb1f9ede244be

File: discussion_grading/enums.py (hunk 12 above)
Comment (BryanttV) [reply discussion_grading/enums.py]: I left it in the main module since it is not used anywhere else.
Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/1ebe6648e62aedfd42acf1d736ba9db8c31c59c8
